File: translation_output/w_test/gpt-3.5/codenet/python/java/p03867.java
```java
import java.util.*;

class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();
        int K = scanner.nextInt();
        final int MOD = 1000000007;
        int i = 1;
        int M = N;
        ArrayList<Integer> div = new ArrayList<>();
        while (i * i <= N) {
            if (N % i == 0) {
                div.add(i);
                if (N / i != i) {
                    div.add(N / i);
                }
            }
            i++;
        }

        Collections.sort(div);
        int N1 = div.size();
        int[] num = new int[N1 + 1];
        int ans = 0;
        for (i = 0; i < N1; i++) {
            num[i] = pow(K, (div.get(i) + 1) / 2, MOD);
            for (int j = 0; j < i; j++) {
                if (div.get(i) % div.get(j) == 0) {
                    num[i] = (num[i] - num[j] + MOD) % MOD;
                }
            }

            if (div.get(i) % 2 == 0) {
                ans += div.get(i) * num[i] / 2;
                ans %= MOD;
            } else {
                ans += div.get(i) * num[i];
                ans %= MOD;
            }
        }
        
        System.out.println(ans);
    }
    
    public static int pow(int a, int b, int MOD) {
        int result = 1;
        while (b > 0) {
            if (b % 2 == 1) {
                result = (result * a) % MOD;
            }
            a = (a * a) % MOD;
            b /= 2;
        }
        return result;
    }
}
```

File: translation_output/w_test/gpt-3.5/codenet/java/python/p03867.py
```python
MOD = 1000000007

def pow(n, k):
    ans = 1
    tmp = n
    while True:
        if k == 0:
            return ans
        if k % 2 == 1:
            ans = (ans * tmp) % MOD
        tmp = (tmp * tmp) % MOD
        k //= 2

def divisors(n):
    lst = []
    for i in range(1, int(n**0.5) + 1):
        if n % i == 0:
            lst.append(i)
            if i != n // i:
                lst.append(n // i)
    lst.sort()
    return lst

def main():
    N, K = map(int, input().split())
    divisor = divisors(N)
    num = [0] * len(divisor)
    tmp_sum = 0
    for i in range(len(divisor)):
        num[i] = pow(K, (divisor[i] + 1) // 2)
        for j in range(i):
            if divisor[i] % divisor[j] == 0:
                num[i] -= num[j]
        while num[i] < 0:
            num[i] += MOD
        tmp_sum += num[i]
        tmp_sum %= MOD

    ans = 0
    for i in range(len(divisor)):
        if num[i] < 0:
            num[i] += MOD
        num[i] = num[i] % MOD
        d = divisor[i]
        if d % 2 == 0:
            ans += num[i] * (d // 2)
        else:
            ans += num[i] * d
        ans %= MOD

    print(ans)

if __name__ == '__main__':
    main()
```

